docs/enhancements/boot-status-reporter.md

```diff
@@ -0,0 +1,129 @@
+   If the boot reporter service fails to run or report logs, the cluster installation should continue, 
+   and the new host progress stages should be skipped in such a scenario.
+
+## Proposal
+
+
+### Changes To Assisted Service
+For cases as mentioned in `Motivation`, we need a new service (start via `systemd`) soon as possible after hosts
+are booting from the disk and contact `assisted-service` to:
+
+1. Change the installation stage to a new `Booted with local ignition` Between `Configuring`
+   and  `Waiting for control plane`. This stage refers to a boot with pointer ignition.
+
+
+2. Change the installation stage to a new `Booted with control plane ignition` Between
+   `Waiting for control plane` and  `Joined`.
+
+
+3. For each of the above-mentioned, Read the current host stage from `assisted-service`, to determine if:
+   - the service runs on the first boot :arrow_right: collect logs and report.
+   - or the second boot :arrow_right: collect logs, report, and self-cleanup after a configurable threshold.
+
+### Information Needed And Changes To The Discovery Agent
+
+In order to communicate with `assisted-service`, The assisted boot reported service needs to know:
+1. The `assisted-service` IP address and port.
+2. The pull-secret to use for agent authentication.
+3. The InfraEnv ID and host ID
+
+It will need this information starting from the first boot, which is a pointer ignition-based boot.
+
+For that, we need to modify the discovery agent to store this information on disk in a pre-defined location for the assisted boot reporter service to use.
+
+The assisted boot reporter service should start via `systemd` and determine its operating stage. As mentioned above:
+
+### First Boot (Pointer Ignition)
+Expected stage to be read from assisted-service backend: `Configuring`.
+Update the host stage to: `Booted with local ignition`.
+Start a timer to compress and send logs every 10 minutes to `v2UpdateHostInstallProgress` API endpoint.
+
+
+### Second Boot (Control Plane Ignition)
+Expected stage to be read from assisted-service backend: `Waiting for control plane`.
+Update the host stage to: `Booted with control plane ignition`.
+Start a timer to compress send logs every 10 minutes to `v2UpdateHostInstallProgress` API endpoint.
+Start another timer to halt operation and self-cleanup as follows:
+
+1. If 1 hour has passed.
+2. If cluster install progress reached a point where it is no longer needed (now counting in `assisted-controller` pod). An indication could be that `kubelet` is up and running.
+   Cleanup should include:
+
+   * The stored files for pull-secret, host id, and infraEnv id.
+   * Service itself - should not run on additional host boot.
+
+### Collected Information
+1. journalctl
+2. A list of mounted file systems
+3. crio logs
+4. network settings
+
+
+### Open Questions
+
+1. What should we use for implementation? Our options are:
```

Comment:
I think we can use both bash and go (container): 
For the boot reporting - use bash, a simple curl command should do the trick, and we don't really on anything other than network access to the service.
For sending logs - I'd prefer if we reuse the send logs functionality already implemented in the assisted-installer-agent. but since this depends on other things (container run time, access to the registry) I guess we can try to run the agent and in case this method fails, fall back to a bash script as well.

Reply:
I don't think we need to rely on pulling image. Idea is to send logs no matter what if we will need to pull image it will break this idea. 
I assume we need to use bash , it should be easy code to write and maintain

Reply:
I tend to agree, the main value of this service is the ability to get the logs from the node with minimum dependencies.
in that regard bash script is the way to go.

Reply:
Why bash? it's not compiled (compilation tells you about mistakes long before you run the program), it's not typed (much easier to write bugs), it's a language riddled with horrible pitfalls and I have no doubt we will be chasing bugs if we use it. 

You can compile a static go binary and place it in the ignition just as easily, then you can use the proper assisted go api library and make structured API calls without relying on curl CLI mess with a real language

Reply:
>The problem with go library is to provide it through ignition.

Why?

>If we decide to go with go library we will need to bring it through pulling image

no images, just a static binary embedded directly inside the ignition

Reply:
For example a statically built controller executable with upx + removing debugging information (`-ldflags="-s -w"`) from the executable it only weighs 20M base64 encoded, and a controller does much more things than what we want to build, and I bet we can make it even smaller if we try

Reply:
I have to disagree here with anything that says "container". We need to ship as a standalone binary at most because we want to catch cases where Container Runtime on the host is broken. No matter whether we would need to pull a new image or just run something that we shipped, if someone thinks about running this service via e.g. `podman run` my very first question will be - what are you going to do if podman binary doesn't execute as you expected?

We had cases where e.g. c/storage was broken. You weren't able to run any container. Not only to pull a container, but not even run what you already have.

If we want to be robust, we should use the simplest stack possible. I agree with Omer on superiority of statically-built go over bash. If we combine it with Nick's idea of just pointing towards the payload instead of embedding it all in the ignition, we will not even increase the ignition's size.

Reply:
+1 @mkowalski.
To download that script, won't that mean we will need to have an endpoint with no auth? (or, "empty auth")?
Won't that be a problem with the SaaS deployment?

Unless.. we can place the pull secret inside the pointer ignition, which allows us to use agent auth.

Reply:
I think we're going to likely have a problem with either release engineering or our build chain if we try to use golang here.

I see a few options and issues with all of them:
1. We compile these binaries from go source and commit them to our code base:
  - Generally poor form. Source repo should have source, not compiled code.
  - I think this will run afoul of some security guidelines that I can't find a reference to at the moment
  - We will need a build chain with all the architectures we support. (with power and Z on the way this could get complicated very quickly)
2. We ship source code to the host and compile there
  - Needs go tooling and dependencies on the host
3. Build an RPM with our changes
  - RPMs are kind of annoying to deal with
  - Need to deal with releng
  - Hard to install on coreos

I see the benefits that @omertuc is referencing, but practically I'm not sure how we would ship something like this to a coreos host without a container image.